closure/layered.go

```diff
@@ -1,130 +1,52 @@
 package closure
 
 import (
-	"sort"
+	"github.com/sirupsen/logrus"
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/simple"
 )
 
-type Node struct {
-	Path         string
-	References   []Node
-	Popularities map[string]int
-}
-
-func buildGraph(storepaths []Storepath) (paths []Node) {
-	roots := findRoots(storepaths)
-	for _, r := range roots {
-		paths = append(paths, buildGraphNode(r, storepaths))
-	}
-	return paths
-}
-
-func findRoots(storepaths []Storepath) (roots []Storepath) {
-	for _, storepath := range storepaths {
-		isRoot := true
-		for _, storepath1 := range storepaths {
-			if storepath.Path == storepath1.Path {
-				continue
-			}
-			for _, ref := range storepath1.References {
-				if storepath.Path == ref {
-					isRoot = false
-					break
-				}
-			}
-		}
-		if isRoot {
-			roots = append(roots, storepath)
-		}
-	}
-	return
-}
-
-func buildGraphNode(storepath Storepath, storepaths []Storepath) (node Node) {
-	node.Path = storepath.Path
-	node.Popularities = make(map[string]int)
-	for _, r := range storepath.References {
-		if r == node.Path {
-			continue
+// SortedPathsByPopularity sorts storepaths by path popularity. It uses the algorithm described in
+// https://grahamc.com/blog/nix-and-layered-docker-images
+func SortedPathsByPopularity(storepaths []Storepath) ([]string, error) {
+	g := simple.NewDirectedGraph()
```

Comment:
I would prefer to move this part in a dedicated function to indicate it builds the graph.

Reply:
doing/done

Reply:
https://github.com/nlewo/nix2container/pull/30/commits/3f335ade6a2ab1e0f8f4a8be5d0f3070a14833b5